rfcs/20190430-tokenization-conventions.md

```diff
@@ -0,0 +1,352 @@
+"""
+Args:
+  input: An N-dimensional `Tensor` or `RaggedTensor` of UTF-8 strings.
+  vocab_lookup_table: A lookup table implementing the LookupInterface
+    containing the vocabulary of subwords.
+  suffix_indicator: (optional) The characters prepended to a wordpiece to
+    indicate that it is a suffix to another subword. Default is '##'.
+  max_bytes_per_word: (optional) Max size of input token. Default is 100.
+  token_out_type: (optional) The type of the token to return. This can be
+    `tf.int64` IDs, or `tf.string` subwords. The default is `tf.int64`.
+  unknown_token: (optional) The value to use when an unknown token is found.
+    Default is "[UNK]". If this is set to a string, and `token_out_type` is
+    `tf.int64`, the `vocab_lookup_table` is used to convert the `unknown_token`
+    to an integer. If this is set to `None`, out-of-vocabulary tokens are left
+    as is.
+
+Returns:
+  A `RaggedTensor`s `tokens` where `tokens[i1...iN, j]` is the string
+        contents, or ID in the vocab_lookup_table representing that string,
+        of the `j`th token in `input[i1...iN]`
+"""
+
+wordpiece_tokenize_with_offsets
+"""
+Args:
+  input: An N-dimensional `Tensor` or `RaggedTensor` of UTF-8 strings.
+  vocab_lookup_table: A lookup table implementing the LookupInterface
+    containing the vocabulary of subwords.
+  suffix_indicator: (optional) The characters prepended to a wordpiece to
+    indicate that it is a suffix to another subword. Default is '##'.
+  max_bytes_per_word: (optional) Max size of input token. Default is 100.
+  token_out_type: (optional) The type of the token to return. This can be
+    `tf.int64` IDs, or `tf.string` subwords. The default is `tf.int64`.
+  unknown_token: (optional) The value to use when an unknown token is found.
+    Default is "[UNK]". If this is set to a string, and `token_out_type` is
+    `tf.int64`, the `vocab_lookup_table` is used to convert the `unknown_token`
+    to an integer. If this is set to `None`, out-of-vocabulary tokens are left
+    as is.
+
+Returns:
+  A tuple of `RaggedTensor`s `tokens`, `start_offsets`, `limit_offsets` where:
+
+    `tokens[i1...iN, j]` is the string contents, or ID in the vocab_lookup_table
+      representing that string, of the `j`th token in `input[i1...iN]`
+    `start_offsets[i1...iN, j]` is the byte offset for the start of the `j`th
+      token in `input[i1...iN]`
+    `limit_offsets[i1...iN, j]` is the byte offset for the end of the `j`th
+      token in `input[i1...iN]`
+"""
+```
+
+#### a custom_tokenizer example {#a-custom_tokenizer-example}
+
+If all tokenizers follow the same principles, it allows for flexibility in
+swapping out tokenization methods, can lend itself to composability, and will be
+easy for anybody already familiar with standard tokenization APIs to use. A
+non-offsets version of a custom tokenize function is provided below that one
+might implement following the same set of guidelines.
+
+```python
+def my_tokenize(input):
```

Comment:
How serious are we with regard to "composability"? The current interface can be quite tricky and burdensome for implementing custom tokenizers using existing tokenizers, especially if offset tracking is required ("Tokenizers should also provide a `_with_offsets` variant" in "Conventions").

For example, consider the extra work needed for implementing my_tokenize_with_offsets().

1. We need to know what input bytes each output code point maps to in every single string rewriting operation, i.e. case_fold_utf8(), regex_replace().
2. For each string splitting operation, we need to translate offsets relative to the splitting input to those relative to the original input.

Reply:
Yeah; I was trying to be careful about the language use of "should" in that it is a best practice, but could be left out if it proved too tricky.

I am updating the rfc to include two interfaces which will make this more explicit.
```
class Tokenizer:
  def tokenize(input):

class TokenizerWithOffsets(Tokenizer):
  def tokenize_with_offsets(input):
```